Add support for _embedded_ PKCS7 signing to signtool_signer sign()

out0xb2 opened the pull request:
To the sign() function:
* added type hints
* In parameter dict "signature_options", the key "sign_alg" is
deprecated and replaced with "format".  Its prior value "pkcs12" was
NOT a signature format, it was the format of the signer's
key file. Moving it to a new signer_option key "key_file_format"
* New "signature_option" key "format" supports 1 value of "pkcs7" which
is the true signature format being produced
* Added "format_options" key to parameter dict "signature_options". Values
"embedded" and "detachedSignatureData" are added to distinguish the
different PKCS7 flavors supported.  Note that these format_options
are mutually exclusive
* To ensure compatibility with existing code that passes
"signature_options" with key "pkcs12", sign() detects this and
fills the new parameters with values the provide backwards compatible
behavior

out0xb2 commented:
All unit tests in edk2toolext\tests\capsule\* pass (with a couple skipped test cases

out0xb2 commented on edk2toolext/capsule/signtool_signer.py:
```diff
@@ -20,6 +20,10 @@
 from edk2toollib.utility_functions import RunCmd
 
 GLOBAL_SIGNTOOL_PATH = None
+SUPPORTED_SIGNTOOL_FORMAT_OPTIONS = {
```
I should probably rename SIGNTOOL to SIGNATURE... as I think this was supposed to be a signature_options['format_options'] ...

out0xb2 replied:
Addressed in commit 2

out0xb2 commented on edk2toolext/capsule/signtool_signer.py:
```diff
@@ -36,18 +40,40 @@ def get_signtool_path():
+    # The following _if_ clause handles the deprecated signature_option 'sign_alg' for backwards compatibility
+    # when the deprecated option is supplied, this code adds the new, required options based on prior code behavior
+    if 'sign_alg' in signature_options:
+        print('!!! WARNING: signature_option "sign_alg" is deprecated, use "format" instead')
```
[Python logging HOWTO](https://docs.python.org/3/howto/logging.html#when-to-use-logging) advises:

> Best tool for the task ...
> warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning
> logging.warning() if there is nothing the client application can do about the situation, but the event should still be noted

corthon replied:
Neat!
I guess I'd go with "warnings.warn".

out0xb2 replied:
Pending commit manual test output:
```
>python -W default::DeprecationWarning create_policy.py

Calling signtool ...
c:\git\edk2-pytool-extensions\edk2toolext\capsule\signtool_signer.py:54: DeprecationWarning: Signature_option "sign_alg" is deprecated, use "format"
  warnings.warn('Signature_option "sign_alg" is deprecated, use "format"', DeprecationWarning)
signtool complete.
```

out0xb2 replied:
Addressed in commit 2

corthon commented on edk2toolext/capsule/pyopenssl_signer.py:
```diff
@@ -21,10 +22,30 @@ def sign(data, signature_options, signer_options):
+        raise ValueError(f"Unsupported signature type: {signature_options['type']}")
+    if 'type_options' in signature_options:
+        raise ValueError(f"Signature type options not supported")
+    if signature_options['encoding'] != 'binary':
```
Binary in this case is PEM?

out0xb2 replied:
As I understand it, PEM is base64 with headers and footers.
This is the raw, binary signed digest... followed by 0x0A... no other headers or footers.

out0xb2 replied:
Scratch to 0x0A, that was being added by an editor.
See line 38 docstring

out0xb2 replied:
Resolved in latest commit

corthon commented on edk2toolext/capsule/pyopenssl_signer.py:
```diff
@@ -21,10 +22,30 @@ def sign(data, signature_options, signer_options):
     dictionaries that are used by capsule_tool and capsule_helper
```
Might as well add the type annotations/hints. ^^^

out0xb2 replied:
Added, tested all configurations, no breaks (of my code anyway).

out0xb2 replied:
Passed unit tests as well, resolved in commit d16b0339e4caaab3d6b7f0bd18027f8b52b34cf2

out0xb2 commented:
> Looks good on the whole. Just added a question and a suggestion.

Updated per your feedback, all unit tests and my test scenario pass (I'm now testing all 4 configuration, OpenSSL, signtool, new parameters, and legacy parameters)

out0xb2 commented:
whoops, not the button I was looking for...

out0xb2 commented:
@corthon , awaiting your, or other admin with write access, to approve
